pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,304 @@
+package cache
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pokt-network/poktroll/pkg/client"
+)
+
+var (
+	_ client.QueryCache[any]           = (*InMemoryCache[any])(nil)
+	_ client.HistoricalQueryCache[any] = (*InMemoryCache[any])(nil)
+
+	DefaultQueryCacheConfig = queryCacheConfig{
+		EvictionPolicy: FirstInFirstOut,
+		// TODO_MAINNET(@bryanchriswhite): Consider how we can "guarantee" good
+		// alignment between the TTL and the block production rate.
+		TTL: time.Minute,
+	}
+)
+
+// InMemoryCache provides a concurrency-safe in-memory cache implementation with
+// optional historical value support.
+type InMemoryCache[T any] struct {
+	config       queryCacheConfig
+	latestHeight atomic.Int64
+
+	itemsMu sync.RWMutex
+	// items type depends on historical mode:
+	// | historical mode |  type                          |
+	// | --------------- | ------------------------------ |
+	// | false           | map[string]cacheItem[T]        |
+	// | true            | map[string]cacheItemHistory[T] |
+	items map[string]any
+}
+
+// cacheItem wraps cached values with a timestamp for later comparison against
+// the configured TTL.
+type cacheItem[T any] struct {
+	value     T
+	timestamp time.Time
+}
+
+// cacheItemHistory stores cachedItems by height and maintains a sorted list of
+// heights for which cached items exist. This list is sorted in descending order
+// to improve performance characteristics by positively correlating index with age.
+type cacheItemHistory[T any] struct {
+	// sortedDescHeights is a list of the heights for which values are cached.
+	// It is sorted in descending order.
+	sortedDescHeights []int64
+	itemsByHeight     map[int64]cacheItem[T]
+}
+
+// NewInMemoryCache creates a new InMemoryCache with the configuration generated
+// by the given option functions.
+func NewInMemoryCache[T any](opts ...QueryCacheOptionFn) *InMemoryCache[T] {
+	config := DefaultQueryCacheConfig
+
+	for _, opt := range opts {
+		opt(&config)
+	}
+
+	return &InMemoryCache[T]{
+		items:  make(map[string]interface{}),
+		config: config,
+	}
+}
+
+// Get retrieves the value from the cache with the given key. If the cache is
+// configured for historical mode, it will return the value at the latest **known**
+// height, which is only updated on calls to SetAtHeight, and therefore is not
+// guaranteed to be the current height.
+func (c *InMemoryCache[T]) Get(key string) (T, error) {
+	if c.config.historical {
+		return c.GetAtHeight(key, c.latestHeight.Load())
+	}
+
+	c.itemsMu.RLock()
+	defer c.itemsMu.RUnlock()
+
+	var zero T
+
+	item, exists := c.items[key]
+	if !exists {
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
+	}
+
+	cItem := item.(cacheItem[T])
+	if c.config.TTL > 0 && time.Since(cItem.timestamp) > c.config.TTL {
+		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
```

Comment:
So is there ever cleanup? Is it something we could defer or do lazyl elsewhere?

Reply:
Yes, will #PUC.

Reply:
```go
	if isTTLEnabled && isCacheItemExpired {
		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
		// otherwise, the read lock would be insufficient. The value will be
		// overwritten by the next call to Set(). If usage is such that values
		// aren't being subsequently set, maxKeys (if configured) will eventually
		// cause the pruning of values with expired TTLs.
		return zero, ErrCacheMiss.Wrapf("key: %s", key)
	}
```

Reply:
I think the `if isTTLEnabled` part was lost as part of another commit.